src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -1320,6 +1346,82 @@ mono_jiterp_is_imethod_var_address_taken (InterpMethod *imethod, int offset) {
 	return mono_bitset_test (imethod->address_taken_bits, offset / MINT_STACK_SLOT_SIZE);
 }
 
+JiterpreterTableInfo tables[JITERPRETER_TABLE_LAST + 1] = { 0 };
+int mono_jiterp_first_trace_fn_ptr = 0;
+
+static void
+atomically_set_value_once (gint32 *address, gint32 value) {
+	gint32 expected = 0;
+	if (atomic_compare_exchange_strong ((atomic_int *)address, &expected, value))
+		return;
+	if (expected == value)
+		return;
+	g_error ("MONO_WASM: Jiterpreter table already initialized with a different value (expected %d, found %d)\n", value, expected);
+}
+
+EMSCRIPTEN_KEEPALIVE void
+mono_jiterp_initialize_table (int type, int first_index, int last_index) {
+	g_assert ((type >= 0) && (type <= JITERPRETER_TABLE_LAST));
+	JiterpreterTableInfo *table = &tables[type];
+	atomically_set_value_once (&table->first_index, first_index);
+	if (type == JITERPRETER_TABLE_TRACE)
+		atomically_set_value_once (&mono_jiterp_first_trace_fn_ptr, first_index);
+	atomically_set_value_once (&table->last_index, last_index);
+
+	// if we lose a race, someone else may have initialized next_index and then bumped it.
+	// that's totally fine though, so don't fail, just ensure it's initialized.
+	gint32 expected = 0;
+	atomic_compare_exchange_strong ((atomic_int *)&table->next_index, &expected, first_index);
+}
+
+EMSCRIPTEN_KEEPALIVE int
+mono_jiterp_allocate_table_entry (int type) {
+	g_assert ((type >= 0) && (type <= JITERPRETER_TABLE_LAST));
+	JiterpreterTableInfo *table = &tables[type];
+	g_assert (table->first_index > 0);
+	// Handle extremely unlikely race condition (allocate_table_entry called while another thread is in initialize_table)
+	gint32 expected = 0;
+	atomic_compare_exchange_strong ((atomic_int *)&table->next_index, &expected, table->first_index);
+	int index = atomic_fetch_add ((atomic_int *)&table->next_index, 1);
+	if (index > table->last_index) {
+		if (index == (table->last_index + 1))
+			g_printf ("MONO_WASM: Jiterpreter table %d is out of space (%d entries allocated)\n", type, index - table->first_index + 1);
```

Comment:
this should probably disable jiterp going forward, right ?

Reply:
No, under normal execution the table could fill up. We don't want to disable the entire jiterpreter because one table is full.